### ISSUE TITLE
[Proposal] add a VSCode 'strict' option, but allow extra features

### ISSUE THREAD

**Issue by mg979** [hrsh7th/vim-vsnip#136]
For example, `let g:vsnip_vscode_strict = 1` to disable all features that are not VSCode compatible. Then you can enable extra features.

Reason: I'd like extra features implemented as keys in the snippet, like:

- `LINE_PATTERN`: only trigger if line matches pattern
- `PATH_PATTERN`: only trigger if file path matches pattern
- `SYNTAX_PATTERN`: only trigger in syntax element with pattern
- `SYNTAX_NO_PATTERN`: don't trigger in syntax elements with pattern

There are snippet that I only want to trigger at the start of an empty line.
Or I have project-specific snippets, that I only want to trigger if the file path matches a certain pattern. Or there are snippets that I don't want to trigger in strings or comments, etc.

Example:
```json
    "NO": {
        "description": "note",
        "body": [
            "[NOTE]",
            "====",
            "$TM_SELECTED_TEXT$0",
            "===="
        ],
        "prefix": [
            "NO"
        ],
        "LINE_PATTERN": "^$"
    },
```

**Comment by hrsh7th** [comment]
What do you think about `vsnip` key?
I think we should separate the plugin-specific configuration to one key.

**Comment by mg979** [comment]
Sounds good.

**Comment by hrsh7th** [comment]
I'm thinking of schema.

**Comment by hrsh7th** [comment]
```
{
  "describe": {
    "prefix": ["describe"],
    "body": [],
    "vsnip": {
      "trigger_pattern": ["^\s*describe"], // auto-expansion
      "bufname_pattern": ["*.test.ts"], // bufname selector
      "syntax_pattern": ["String"] // syntax selector
    }
  }
}
```

**Comment by hrsh7th** [comment]
```
{
  "describe": {
    "prefix": ["describe"],
    "body": [
      "..."
    ],
    "vsnip": {
      "syntax_not": ["Comment", "String"],
      "bufname": "\\.test\\.[jt]sx?",
      "auto_expand": "desc(",
    }
  }
}
```

The `syntax` or `bufname` are `conditional` options.
The conditional options can be having a suffix `_not`.

In what order should the options be applied?

**Comment by mg979** [comment]
I don't think it matters because all conditions should be satisfied, so snippet is expanded only if all conditions are true.

**Comment by gzagatti** [comment]
I am wondering if there is any updates on this proposal. It would be great to have the ability to only trigger some snippets according to these line patterns.
